Draft of abort_timeout and redis namespace

dp-alvarez opened the pull request:
### Backend
- Keys are now strings
- Keys are stored with a dict containing parameters to the abort request
- Redis takes an optional namespace to store keys on

### Middleware
- Abort includes an abort_timeout parameter
- Tasks can check if an abort is requested via `abort_requested`. Returns the number of seconds the task has before an `Abort` exception is raised or `None` if no abort is requested
- "Main loop" split into fetching abort messages and aborting pending tasks

### TODO
- Gevent abort manager
- Tests
- Update documentation

isra17 commented on src/dramatiq_abort/backend.py:
```diff
@@ -1,14 +1,14 @@
-        self, keys: List[bytes], timeout: int
-    ) -> Optional[bytes]:  # pragma: no cover
+        self, keys: Iterable[str], timeout: int
+    ) -> Tuple[Optional[str], Optional[Dict[str, Any]]]:  # pragma: no cover
```
I would rather use `NamedTuple` if the tuple components are not obvious. Also, Let's make the tuple optional, not its component.

```python
class Event(NamedTuple):
  type: str
  params: Dict[str, Any]

def wait_many(...) -> Optional[Event]: ...
def poll(...) -> Optional[Event]: ...
def notify(self, items: Iterable[Event], ...): ...
```
```

isra17 commented on src/dramatiq_abort/backends/redis.py:
```diff
@@ -14,38 +18,97 @@ class RedisBackend(EventBackend):
     """
 
-    def __init__(self, *, client: Any) -> None:
+    def __init__(self, *, client: Any, namespace: Optional[str] = None) -> None:
```
No need for Optional, just default to `DEFAULT_NAMESPACE`. `str` are immutable, so safe to share as default.

isra17 commented on src/dramatiq_abort/backends/redis.py:
```diff
@@ -14,38 +18,97 @@ class RedisBackend(EventBackend):
                 pipe.pexpire(key, ttl)
             pipe.execute()
+
+    def _encode(
```
I would rather split `_encode` into `_encode_key` and `_encode_value` to have:

```python
rpush(self._encode_key(event.key), self._encode_value(event.params))
lpop(self._encode_key(key))
```

isra17 replied:
And do the same for `_decode_*`

isra17 commented on src/dramatiq_abort/backends/redis.py:
```diff
@@ -1,9 +1,13 @@
 
 from ..backend import EventBackend
 
+MARKER = "dramatiq-abort_marker"
```
to be fair, I don't think the `MARKER` is worth to keep with namespace. I would remove this.

isra17 commented on src/dramatiq_abort/middleware.py:
```diff
@@ -144,53 +144,74 @@ def abort(
-    def _handle(self) -> None:
-        message_ids = list(self.manager.abortables.keys())
+    def _get_abort_requests(self) -> None:
+        message_ids = self.manager.abortable_messages.keys()
```
why did you remove the `list`? It's required to avoid iterating on mutated .dictionnary

dp-alvarez replied:
Nice catch! I had edited that code and forgot the list when I changed it back

isra17 commented on src/dramatiq_abort/middleware.py:
```diff
@@ -144,53 +144,74 @@ def abort(
     middleware: Optional[Abortable] = None,
     abort_ttl: Optional[int] = None,
     mode: AbortMode = AbortMode.ABORT,
+    abort_timeout: int = 0,
```
Please update the docstring

isra17 commented on src/dramatiq_abort/middleware.py:
```diff
@@ -231,24 +253,54 @@ class _CtypesAbortManager:
+                message_id, abort_time = self.abort_requests.pop(thread_id, ("", None))
+                saved_thread_id = self.abortable_messages.pop(message_id, None)
+                assert saved_thread_id == thread_id
+                if thread_id is None or abort_time is None:
```
I don't see how this condition could ever be `True`. We are inside the lock, so `abort_request[thread_id]` is guaranteed to have value.

isra17 replied:
Actually I'm pretty sure that we really only need line 300-303

dp-alvarez replied:
That is true, the `if` really should be removed.

Although we technically only need the last few lines I think it is better to keep the pops. If for whatever reason remove_abortable doesn't run we would end up aborting the wrong task, so remove them from the list just in case.

isra17 reviewed (changes requested):
Few comments and improvements! Overall look good to me, thanks for the work!

dp-alvarez commented:
Awesome! I will send in another PR once I finish the docs and tests.

dp-alvarez commented:
I extracted AbortManager into a abstract class and added support for gevent. Let me know if it looks ok.

Docstrings, tests and documentation still todo.

isra17 commented on src/dramatiq_abort/abort_manager.py:
```diff
@@ -0,0 +1,142 @@
+
+    @abc.abstractmethod
+    def __init__(self, logger: Optional[Logger] = None):
+        self.logger = logger or get_logger(__name__, type(self))
```
I think we might as well set `abortable_messages` and `abort_requests` here since its quite core to the base.

`lock` could also be defined as an abstract property, forcing implementer to define it, ie:

```
@abc.abstractmethod
@property
def lock(self) -> ContextManager[Any]: ...
  # CTypesManager impl: return self._lock
```

isra17 commented on src/dramatiq_abort/abort_manager.py:
```diff
@@ -0,0 +1,142 @@
+    logger: Any
+    lock: ContextManager[Any]
+    abortable_messages: Dict[str, Any]
+    abort_requests: Dict[Any, Tuple[str, float]]
```
Can we also use a `AbortRequest` NamedTuple here?

isra17 commented on src/dramatiq_abort/abort_manager.py:
```diff
@@ -0,0 +1,142 @@
+            thread = self.abortable_messages.get(message_id, None)
+        else:
+            thread = self.get_current_thread()
+        _, abort_time = self.abort_requests.get(thread, (None, None))
```
I would rather keep the longer, but simpler alternative:

```python
abort_request = self.abort_requests.get(thread)
if not abort_request:
  return None
return abort_request.abort_time # assuming namedtuple
```

isra17 commented on src/dramatiq_abort/abort_manager.py:
```diff
@@ -0,0 +1,142 @@
+            ]
+
+            for thread in toabort:
+                message_id, abort_time = self.abort_requests.pop(thread, ("", None))
```
as above, not a fan of artificial defaults, prefer:

```python
abort_request = self.abort_requests.pop(thread, None)
if abort_request:
  self.abortable_messages.pop(abort_request.key, None)
```

dp-alvarez commented:
Just finished updating the docs, tests and requested changes (along with some misc fixes along the way).
Appreciate the review! Certainly made the code much cleaner.

Let me know if it needs any tweaks, or feel free to change anything as you see fit.

isra17 commented:
I merged by hand (And fixed the CI), thanks a lot again!

dp-alvarez commented:
I see a new version was released too, great.

And thank you for this project! Saved me a lot of time compared to doing it from scratch.
